### PR TITLE
Database changes observe with Flow, RecyclerView adapter change and Optimization

### PR DESCRIPTION
**Pull request by jaypatel9912**
Database changes observe with Flow
new way adapter used to show data in RecyclerView (as in dcl)
code optimization

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -50,7 +50,8 @@ dependencies {
     implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.3.1"
     implementation "androidx.lifecycle:lifecycle-runtime-ktx:2.4.0-alpha03"
     implementation "androidx.fragment:fragment-ktx:1.3.6"
+    implementation group: 'com.google.guava', name: 'guava', version: '24.1-jre'
```
Comment (rkrsgithub) [inline app/build.gradle]: Why we need this unwanted library?

File: app/src/main/java/com/demoapp/adapterutils/BaseAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
+package com.demoapp.adapterutils
+
+import android.content.Context
+
+/**
+ * Recycler adapter used to display the different More screen sections.
+ * @author roger.cotrina, VELEJ097
+ */
+open class BaseAdapter(map: HashMap<Int, DelegateAdapter<*, *>>, val context: Context) :
```
Comment (rkrsgithub) [inline app/src/main/java/com/demoapp/adapterutils/BaseAdapter.kt]:
```suggestion
open class BaseAdapter(map: HashMap<Int, DelegateAdapter<*, *>>) :
```

File: app/src/main/java/com/demoapp/database/DataRepository.kt [hunk 3]
```diff
@@ -1,71 +1,73 @@
...
-            response.forEach { category ->
-                dataDuo.insertCategories(CategoryEntity(name = category.category))
+            dataDuo.insertAllCategories(response.map {
+                CategoryEntity(name = it.category)
+            })
 
-                category.items.forEach {
-                    dataDuo.insertData(
+            response.map {
+                dataDuo.insertAllSubCategories(
```
Comment (rkrsgithub) [inline app/src/main/java/com/demoapp/database/DataRepository.kt]: Why we need insert inside map ?

File: app/src/main/java/com/demoapp/model/DataEntity.kt [hunk 4]
```diff
@@ -6,12 +6,11 @@ import androidx.room.PrimaryKey
...
-    @ColumnInfo(name = "product_id") val product_id: String,
-    @ColumnInfo(name = "category_name") val category_name: String,
-    @ColumnInfo(name = "is_category") val is_category: Boolean,
+    @PrimaryKey @ColumnInfo(name = "name") val name: String = "",
+    @ColumnInfo(name = "description") val description: String = "",
+    @ColumnInfo(name = "price") val price: Double = 0.0,
+    @ColumnInfo(name = "image_link") val image_link: String = "",
+    @ColumnInfo(name = "product_id") val product_id: String = "",
+    @ColumnInfo(name = "category_name") val category_name: String = "",
+    @ColumnInfo(name = "is_category") val is_category: Boolean = false,
```
Comment (rkrsgithub) [inline app/src/main/java/com/demoapp/model/DataEntity.kt]:
```suggestion
    @ColumnInfo(name = "image_link") val imageLink: String = "",
    @ColumnInfo(name = "product_id") val productId: String = "",
    @ColumnInfo(name = "category_name") val categoryName: String = "",
    @ColumnInfo(name = "is_category") val isCategory: Boolean = false,
```

File: app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt [hunk 5]
```diff
@@ -2,21 +2,21 @@ package com.demoapp.viewmodel
...
 
 class MainActivityViewModel @Inject constructor(private val dataRepository: DataRepository) :
     ViewModel() {
 
-    private var _allCategoryList = MutableStateFlow<List<DataEntity>>(emptyList())
     private var _showLoader = MutableStateFlow(false)
     private var _message = MutableStateFlow("")
 
-    val categoryList: Flow<List<DataEntity>> = _allCategoryList
+    var categoryList: Flow<List<RecyclerViewType>> = dataRepository.getLocalData()
```
Comment (rkrsgithub) [inline app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt]:
```suggestion
    val categoryList: Flow<List<RecyclerViewType>> = dataRepository.getLocalData()
```

File: app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt [hunk 6]
```diff
@@ -25,22 +25,26 @@ class MainActivityViewModel @Inject constructor(private val dataRepository: Data
...
 
-    private fun populateData() {
+    private fun getDataFromApi() {
         viewModelScope.launch {
-            val resultList = dataRepository.getLocalData()
-            _allCategoryList.value = resultList
+            val res = dataRepository.getCategoryData()
+            if (res != null) {
+                dataRepository.insertData(res)
+            }
```
Comment (rkrsgithub) [inline app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt]: do this triggers from Repository itself

**Review by rkrsgithub — CHANGES_REQUESTED** [review]

File: app/build.gradle (hunk 1 above)
Comment (jaypatel9912) [reply app/build.gradle]: Raghu, classes from this library are used in CollectionsUtils and GenericBaseRecyclerViewAdapter Files

import com.google.common.base.Preconditions
import com.google.common.collect.ImmutableMap.Builder

import com.google.common.collect.Lists

File: app/src/main/java/com/demoapp/adapterutils/BaseAdapter.kt (hunk 2 above)
Comment (jaypatel9912) [reply app/src/main/java/com/demoapp/adapterutils/BaseAdapter.kt]: Removed

File: app/src/main/java/com/demoapp/model/DataEntity.kt (hunk 4 above)
Comment (jaypatel9912) [reply app/src/main/java/com/demoapp/model/DataEntity.kt]: variable name changed

File: app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt (hunk 5 above)
Comment (jaypatel9912) [reply app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt]: Changed var to val

File: app/src/main/java/com/demoapp/database/DataRepository.kt (hunk 3 above)
Comment (jaypatel9912) [reply app/src/main/java/com/demoapp/database/DataRepository.kt]: Using map, app creates model to be inserted to local db from api response
Creating **CategoryEntity** from **category** field in response
Creating **DataEntity** from response with **categoryName** to manage reference between 2 tables

File: app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt (hunk 6 above)
Comment (jaypatel9912) [reply app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt]: Created DataSource for transactions with local database

File: app/src/main/java/com/demoapp/database/DataRepository.kt (hunk 3 above)
Comment (jaypatel9912) [reply app/src/main/java/com/demoapp/database/DataRepository.kt]: Removed this part and managed through data source

File: app/build.gradle (hunk 1 above)
Comment (rkrsgithub) [reply app/build.gradle]: I think we are not using that things. keep only needed things

Comment (jaypatel9912) [reply app/build.gradle]: Removed library

**Review by rkrsgithub — APPROVED** [review]

**Pull request merged**